File: lincoln/ruby/lib/lang/plan/predicate.rb
require 'lib/lang/plan/term'
require 'lib/lang/plan/dont_care'
require 'lib/lang/plan/arguments'
require 'lib/types/operator/index_join'
require 'lib/types/operator/scan_join'
require 'lib/types/operator/anti_scan_join'
require 'lib/lang/plan/object_from_catalog'
require 'lib/lang/parse/schema.rb'

class Predicate < Term 
  include Enumerable
	
	def initialize(notin, name, event, arguments) 
		super()
		@notin = notin
		@name = name
		@event = event
		@arguments = Arguments.new(self, arguments)
	end


	class Event
	  NONE = 0
	  INSERT = 1 
	  DELETE = 2
  end

	attr_reader :schema, :notin, :name
	attr_accessor :event
	
	def locationVariable
		self.each do |argument|
			return argument if argument.class <= Variable and argument.loc
		end
		nil
	end
	
	def containsAggregation
		@arguments.each do |e|
			if e.class <= Aggregate then
				return true;
		  end
	  end
		
		false
	end

	# An iterator over the predicate arguments.
  def each
    @arguments.each do |a|
      yield a
    end
  end
	
	def argument(i)
		return @arguments[i]
	end
	
	# was called "arguments" in the Java code!!
	def num_arguments
		return @arguments.length
	end
	
  def to_s
		value = (notin ? "notin " : "") + name.to_s + "("
		if @arguments.length == 0
			return value + ")"
		end
		value += argument(0).to_s
		(1..@arguments.length-1).each do |i|
			value += ", " + argument(i).to_s
		end
		value += argument(@arguments.length).to_s
		return value + ")"
	end

	def requires
		variables = Hash.new
		@arguments.each do |a|
			if (!(a.class <= Variable)) then
			  a.variables.each{|v| variables << v}
			end
		end
		return variables
	end

  def operator(context, input) 
		# Determine the join and lookup keys.
		lookupKey = Key.new
		indexKey  = Key.new
		@schema.variables.each do |var|
			if (input.contains(var)) then
				indexKey << var.position
				lookupKey << input.variable(var.name).position
			end
		end
		
		if (@notin) then
			return AntiScanJoin.new(context, self, input)
		end
		
		table = context.catalog.table(@name)
		index = nil
		if (indexKey.size > 0) then
			if (table.primary.key == indexKey) then
				index = table.primary
			elsif (table.secondary.has_key?(indexKey))
				index = table.secondary[indexKey]
			else
				index = HashIndex.new(context, table, indexKey, Index::Type::SECONDARY)
				table.secondary[indexKey] = index
			end
		end
		
		if !index.nil? then
			return IndexJoin.new(context, self, input, lookupKey, index)
		else
			return ScanJoin.new(context, self, input)
		end
	end
	
	def set(context, program, rule, position) 
	  super(context,program,rule,position)
#    Program.predicate.force(Tuple.new(program, rule, position, event.to_s, self))
    ttup = context.catalog.table(PredicateTable.table_name)
    raise "table " + PredicateTable.table_name.to_s + " missing from catalog!" if ttup.nil?
    ttup.force(Tuple.new(program, rule, position, event, self))
		@schema = Schema.new(name, nil)
		@arguments.each do |arg|
			if (arg.class <= Variable) then
				@schema << arg
			else 
				@schema << DontCare.new(arg.class, position, nil)
			end
		end
	end
	
end

File: lincoln/java/trunk/src/jol/lang/plan/Predicate.java
package jol.lang.plan;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Set;

import jol.core.Runtime;
import jol.types.basic.Schema;
import jol.types.basic.Tuple;
import jol.types.exception.PlannerException;
import jol.types.exception.UpdateException;
import jol.types.operator.AntiIndexJoin;
import jol.types.operator.AntiScanJoin;
import jol.types.operator.IndexJoin;
import jol.types.operator.Operator;
import jol.types.operator.ScanJoin;
import jol.types.table.HashIndex;
import jol.types.table.Index;
import jol.types.table.Key;
import jol.types.table.ObjectTable;
import jol.types.table.Table;
import jol.types.table.TableName;

public class Predicate extends Term implements Iterable<Expression> {
	public enum Field{PROGRAM, RULE, POSITION, EVENT, OBJECT};
	public static class PredicateTable extends ObjectTable {
		public static final TableName TABLENAME = new TableName(GLOBALSCOPE, "predicate");
		public static final Key PRIMARY_KEY = new Key(0,1,2);

		public static final Class[] SCHEMA =  {
			String.class,     // program name
			String.class,     // rule name
			Integer.class,    // position
			String.class,     // Event
			Predicate.class   // predicate object
		};

		public PredicateTable(Runtime context) {
			super(context, TABLENAME, PRIMARY_KEY, SCHEMA);
		}

		@Override
		protected boolean insert(Tuple tuple) throws UpdateException {
			Predicate object = (Predicate) tuple.value(Field.OBJECT.ordinal());
			if (object == null) {
				throw new UpdateException("Predicate object null");
			}
			object.program   = (String) tuple.value(Field.PROGRAM.ordinal());
			object.rule      = (String) tuple.value(Field.RULE.ordinal());
			object.position  = (Integer) tuple.value(Field.POSITION.ordinal());
			return super.insert(tuple);
		}

		@Override
		protected boolean delete(Tuple tuple) throws UpdateException {
			return super.delete(tuple);
		}
	}

	public enum Event{NONE, INSERT, DELETE};

	public Runtime context;

	private boolean notin;

	private TableName name;

	private Event event;

	private Arguments arguments;

	public Predicate(Runtime context, boolean notin, TableName name, Event event, List<Expression> arguments) {
		super();
		this.context = context;
		this.notin = notin;
		this.name = name;
		this.event = event;
		this.arguments = new Arguments(this, arguments);
	}

	public Predicate(Runtime context, boolean notin, TableName name, Event event, Schema schema) {
		super();
		this.context = context;
		this.notin = notin;
		this.name = name;
		this.event = event;
		this.arguments = new Arguments(this, schema.variables());
	}

	protected Predicate(Predicate pred) {
		super();
		this.context   = pred.context;
		this.notin     = pred.notin;
		this.name      = pred.name;
		this.event     = pred.event;
		this.arguments = pred.arguments != null ? pred.arguments.clone() : null;
	}

	public Predicate clone() {
		return new Predicate(this);
	}

	public Schema schema() {
		List<Variable> variables = new ArrayList<Variable>();
		for (Expression e : this.arguments) {
			if (e instanceof Variable) {
				variables.add((Variable) e);
			}
			else {
				variables.add(new DontCare(e.node(), e.type()));
			}
		}
		return new Schema(this.name, variables);
	}
	
	@Override
	public Schema schema(Schema input) {
		if (!this.notin) {
			for (Variable var : this.schema().variables()) {
				if (!input.contains(var)) {
					input.append(var);
				}
			}
		}
		return input;
	}

	public boolean notin() {
		return this.notin;
	}

	public Event event() {
		return this.event;
	}

	void event(Event event) {
		this.event = event;
	}

	public TableName name() {
		return this.name;
	}

	public Variable locationVariable() {
		for (Expression argument : this) {
			if (argument instanceof Variable && ((Variable) argument).loc()) {
				return (Variable) argument;
			}
		}
		return null;
	}

	public boolean containsAggregation() {
		for (Expression e : arguments) {
			if (e instanceof Aggregate) {
				return true;
			}
		}
		return false;
	}

	/**
	 * An iterator over the predicate arguments.
	 */
	public Iterator<Expression> iterator() {
		return this.arguments.iterator();
	}

	public Expression argument(Integer i) {
		return this.arguments.get(i);
	}
	
	public void remove(Expression e) {
		this.arguments.remove(e);
	}

	public List<Expression> arguments() {
		return this.arguments;
	}

	void arguments(List<Expression> arguments) {
		this.arguments = new Arguments(this, arguments);
	}

	@Override
	public String toString() {
		String value = (notin ? "notin " : "") + name + "(";
		if (arguments.size() == 0) {
			return value + ")";
		}
		value += arguments.get(0).toString();
		for (int i = 1; i < arguments.size(); i++) {
			value += ", " + arguments.get(i);
		}
		return value + ")";
	}

	@Override
	public Set<Variable> requires() {
		Set<Variable> variables = new HashSet<Variable>();
		for (Expression<?> arg : arguments) {
			if (!(arg instanceof Variable)) {
				variables.addAll(arg.variables());
			}
		}
		return variables;
	}

	@Override
	public Operator operator(Runtime context, Schema input) throws PlannerException {
		/* Determine the join and lookup keys. */
		Key lookupKey = new Key();
		Key indexKey  = new Key();
		Schema schema = schema();
		for (Variable var : schema.variables()) {
			if (input.contains(var)) {
				indexKey.add(schema.position(var.name()));
				lookupKey.add(input.position(var.name()));
			}
		}

		Table table = context.catalog().table(this.name);
		Index index = null;
		if (indexKey.size() > 0) {
			if (table.primary().key().equals(indexKey)) {
				index = table.primary();
			}
			else if (table.secondary().containsKey(indexKey)) {
				index = table.secondary().get(indexKey);
			}
			else {
				index = new HashIndex(context, table, indexKey, Index.Type.SECONDARY);
				table.secondary().put(indexKey, index);
			}
		}
		
		if (notin) {
			if (index != null) {
				return new AntiIndexJoin(context, this, input, lookupKey, index);
			} else {
				return new AntiScanJoin(context, this, input);
			}
		} else {
			if (index != null) {
				return new IndexJoin(context, this, input, lookupKey, index);
			} else {
				return new ScanJoin(context, this, input);
			}
		}
	}

	@Override
	public void set(Runtime context, String program, String rule, Integer position) throws UpdateException {
		context.catalog().table(PredicateTable.TABLENAME)
			.force(new Tuple(program, rule, position, event.toString(), this));
	}

}
